support binarize for ThresholdAlgorithm

johnnychen94 opened the pull request:
Introduce a non-exported trait type `BinaryHistogramThreshold` to wrap `ThresholdAlgorithm`. Except for the change of type hierarchy, there are no other behavior changes.


* To make the review easier, I didn't remove the previous codes. Will remove that later.
* Documentations can be added later.

closes #67

johnnychen94 committed:
support binarize for ThresholdAlgorithm

johnnychen94 commented on src/algorithms/binary_histogram_threshold.jl:
```diff
@@ -0,0 +1,36 @@
+struct BinaryHistogramThreshold{T} <: AbstractImageBinarizationAlgorithm
```
`BinaryHistogramThreshold` seems not a good name, any suggestions?

zygmuntszpak replied:
How about `ImageHistogramThreshold` ?

johnnychen94 replied:
The "Binary" here indicates that it's a method that uses only one threshold value, which is inferred from whatever histogram thresholding methods.

zygmuntszpak replied:
What about 
```julia
struct SingleHistogramThreshold{T} <: AbstractImageBinarizationAlgorithm
```

johnnychen94 commented:
@zygmuntszpak if you're good on this, I can continue the clean-up work here.

zygmuntszpak commented:
> @zygmuntszpak if you're good on this, I can continue the clean-up work here.
This is fantastic, thank you for all your time and effort.

johnnychen94 committed:
add `nbins` keywords and tweak the performance

Also add docstrings and tests for the new type. Docstrings and tests for
the old types are absorbed into the new ones and hence no longer needed.

johnnychen94 committed:
remove unused codes

johnnychen94 commented:
There are other two changes:

1. The broadcasting semantic is more performant than a manually written loop:

```julia
julia> @btime binarize($img, Otsu());
  161.357 μs (10 allocations: 273.27 KiB) # this branch
  386.131 μs (10 allocations: 273.27 KiB) # master
```

2. I added a new keyword `nbins` to `binarize` for `ThresholdAlgorithm`, e.g.,:

```julia
binarize(img, Otsu(); nbins=32)
```

Tests for those `ThresholdAlgorithms` are refactored in a more compact for-loop with some new tests for the new usage.

Except for the docs, I think this PR is mostly done.

johnnychen94 commented on test/algorithms/single_histogram_threshold.jl:
```diff
@@ -0,0 +1,101 @@
+            for T in type_list
+                img = T.(img_gray)
+                reffile = joinpath("References", fname*"_Gray.png")
+                @test_reference reffile Gray.(binarize(img, f)) by=binarization_equality()
```
We're still testing against the previous references, so yeah this PR is pretty safe except for the type hierarchy changes.

johnnychen94 commented on src/ImageBinarization.jl:
```diff
@@ -47,17 +41,11 @@ export
-    Yen
+    
+    # also reexport algorithms in HistogramThresholding
+    SingleHistogramThreshold
```
The only reason I export this is because I want to make `? SingleHistogramThreshold` work.

Since I plan to introduce this package into Images.jl, we might be conservative and not export this?

johnnychen94 replied:
I still choose to export this for completeness; people might wonder how `Otsu` and such get called.

johnnychen94 commented on src/algorithms/single_histogram_threshold.jl:
```diff
@@ -0,0 +1,107 @@
+
+## `alg::ThresholdAlgorithm`
+
+`ThresholdAlgorithm` is an Abstract type defined in `ThresholdAlgorithm.jl`, it provides various
+threshold finding algorithms:
+
+$(mapreduce(x->"- [`"*string(x)*"`](@ref)", (x,y)->x*"\n"*y, threshold_methods))
```
preview for this:

```markdown
`ThresholdAlgorithm` is an Abstract type defined in `ThresholdAlgorithm.jl`, it provides various threshold finding algorithms:

- [`HistogramThresholding.Balanced`](@ref)
- [`HistogramThresholding.Entropy`](@ref)
- ...
- [`HistogramThresholding.Yen`](@ref)
```

johnnychen94 committed:
more tweak to performance

zygmuntszpak reviewed (approved):
Looks great, thank you for all your time and effort!
